server/lib/mutex.ts

```diff
@@ -44,3 +44,40 @@ export async function lockUntilResolved<T>(
     debug(`Released lock ${_key}`);
   });
 }
+
+/**
+ * Try to atomically acquire a lock for a key, execute the `until` function and release the lock.
+ * Throws if the lock can't be acquired.
+ *
+ * Requires a Redis instance to be available, otherwise it will just run the `until` function.
+ */
+export async function lockUntilOrThrow<T>(
+  key: string,
+  until: () => Promise<T>,
+  { unlockTimeoutMs = 10 * 60 * 1000 } = {},
+): Promise<T> {
+  const redis = await createRedisClient(RedisInstanceType.SESSION);
+  if (!redis) {
+    logger.warn(`Redis is not available, ${key} running without a mutex lock!`);
+    return until();
+  }
+
+  const _key = `lock:${key}`;
+  const lockAcquired = await redis.set(_key, 1, { NX: true, PX: unlockTimeoutMs });
+  if (!lockAcquired) {
+    debug(`Failed to acquire lock ${_key}`);
+    throw new Error(`Failed to acquire lock for key ${_key}`);
+  }
+
+  debug(`Acquired lock ${_key}`);
+  try {
+    const response = await until();
+    await redis.del(_key);
+    debug(`Released lock ${_key}`);
+    return response;
+  } catch (e) {
+    await redis.del(_key);
+    debug(`Released lock ${_key}`);
+    throw e;
+  }
```

Comment:
This code flow looks a bit strange, `catch` could be triggered because Redis is unavailable. If the goal is just to clear the lock whatever happens, maybe you intended to do something like this?

```suggestion
  try {
    return await until();
  } finally {
    await redis.del(_key);
    debug(`Released lock ${_key}`);
  }
```

Reply:
The issue here is that the finally handler would eat any error thrown by until().

Reply:
It won't. If an error is thrown by `until`, `finally` will run then the error will bumble up. Try this:

```ts
const testFinally = async (fn) => {
  try {
    return await fn();
  } finally {
    console.log('finally');
  }
}

const test = async () => {
  console.log('return value with no error:', await testFinally(() => Promise.resolve('hello')));
  
  console.log('Testing error:')
  try {
    await testFinally(() => Promise.reject('error'))
  } catch (e) {
    console.log('received error:', e);
  }
}


test()
```

> $ test()
> finally
> return value with no error: hello
> Testing error:
> finally
> received error: error

Reply:
The key here is `return await`, that's how we make sure the error is raised inside of this `try finally`